validator/c/src/lib.rs

```diff
@@ -0,0 +1,609 @@
+/// descriptions, but haven't yet been implemented in the validator) if the
+/// fields are set to their default value. If this option isn't set, or if an
+/// unknown field is not set to its default value, a warning is emitted.
+///
+/// Returns whether the function was successful. If false is returned, retrieve
+/// the error message with substrait_validator_get_last_error().
+#[no_mangle]
+pub extern "C" fn substrait_validator_config_ignore_unknown_fields_set_to_default(
+    config: *mut ConfigHandle,
+) -> bool {
+    // Unpack configuration handle, catching null pointers.
+    if config.is_null() {
+        set_last_error("received null configuration handle");
+        return false;
+    }
+    let config = unsafe { &mut (*config).config };
+
+    // Update configuration and return success.
+    config.ignore_unknown_fields_set_to_default();
+    true
+}
+
+/// Whitelists a protobuf message type for use in advanced extensions. If an
+/// advanced extension is encountered that isn't whitelisted, a warning is
+/// emitted. The type URL pattern may include * and ? wildcards for glob-like
+/// matching (see https://docs.rs/glob/latest/glob/struct.Pattern.html for the
+/// complete syntax).
+///
+/// Returns whether the function was successful. If false is returned, retrieve
+/// the error message with substrait_validator_get_last_error().
+#[no_mangle]
+pub extern "C" fn substrait_validator_config_whitelist_any_url(
+    config: *mut ConfigHandle,
+    pattern: *const libc::c_char,
+) -> bool {
+    // Unpack configuration handle, catching null pointers.
+    if config.is_null() {
+        set_last_error("received null configuration handle");
+        return false;
+    }
+    let config = unsafe { &mut (*config).config };
+
+    // Unpack URL, catching null pointers.
+    if pattern.is_null() {
+        set_last_error("received null pattern");
+        return false;
+    }
+    let pattern = unsafe { std::ffi::CStr::from_ptr(pattern) };
+    let pattern = match pattern.to_str() {
+        Ok(u) => u,
+        Err(e) => {
+            set_last_error(format!("received invalid pattern: {}", e));
+            return false;
+        }
+    };
+    let pattern = match substrait_validator_core::Pattern::new(pattern) {
+        Ok(p) => p,
+        Err(e) => {
+            set_last_error(format!("received invalid pattern: {}", e));
+            return false;
+        }
+    };
+
+    // Update configuration and return success.
+    config.whitelist_any_url(pattern);
+    true
+}
+
+/// Converts a positive/zero/negative integer into Info/Warning/Error
+/// respectively.
+fn int_to_level(x: i32) -> substrait_validator_core::Level {
+    match x.cmp(&0) {
+        std::cmp::Ordering::Greater => substrait_validator_core::Level::Info,
+        std::cmp::Ordering::Equal => substrait_validator_core::Level::Warning,
+        std::cmp::Ordering::Less => substrait_validator_core::Level::Error,
+    }
```

Comment:
I find something like this easier to understand/read: 
```suggestion
    match x {
        1..=i32::MAX => substrait_validator_core::Level::Info,
        0 => substrait_validator_core::Level::Warning,
        i32::MIN..=-1 => substrait_validator_core::Level::Error,
    }
```